synapse/storage/databases/main/events_worker.py

```diff
@@ -1055,23 +1056,46 @@ async def have_seen_events(self, event_ids):
         Returns:
             set[str]: The events we have already seen.
         """
+        res = await self._have_seen_events_dict(event_ids)
+        return {x for (x, y) in res.items() if y}
+
+    @cachedList("have_seen_event", "event_ids")
+    async def _have_seen_events_dict(
+        self, event_ids: Collection[str]
+    ) -> Dict[str, bool]:
+        """Helper for have_seen_events
+
+        Returns a dict, which is the right format for @cachedList
+        """
         # if the event cache contains the event, obviously we've seen it.
-        results = {x for x in event_ids if self._get_event_cache.contains(x)}
+        cache_results = {x for x in event_ids if self._get_event_cache.contains(x)}
+        results = {x: True for x in cache_results}
 
         def have_seen_events_txn(txn, chunk):
+            # assume everything in this chunk is not found initially
+            results.update({x: False for x in chunk})
+
+            # check the db and update the results for any row that is found
             sql = "SELECT event_id FROM events as e WHERE "
             clause, args = make_in_list_sql_clause(
                 txn.database_engine, "e.event_id", chunk
             )
             txn.execute(sql + clause, args)
-            results.update(row[0] for row in txn)
+            results.update({row[0]: True for row in txn})
 
-        for chunk in batch_iter((x for x in event_ids if x not in results), 100):
+        for chunk in batch_iter((x for x in event_ids if x not in cache_results), 100):
             await self.db_pool.runInteraction(
                 "have_seen_events", have_seen_events_txn, chunk
             )
+
         return results
 
+    @cached(max_entries=100000)
```

Comment:
is this a sensible size? The entries are pretty tiny.

Reply:
I think that's fine.